Add ro.go.hmlendea.DL-Desktop

hmlendea opened the pull request:
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/hmlendea/duolingo-desktop
- [X] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

hfiguiere commented on ro.go.hmlendea.Duolingo.desktop:
```diff
@@ -0,0 +1,12 @@
+Type=Application
+Comment=Learn new languages in bite-sized lessons
+Comment[ro]=Învață limbi noi în lecții mărunte
+Exec=
```
you need an exec line

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+app-id: ro.go.hmlendea.Duolingo
+branch: stable
```
```suggestion
```

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+runtime-version: '20.08'
+sdk: org.freedesktop.Sdk
+base: org.electronjs.Electron2.BaseApp
+base-version: '20.08'
```
```suggestion
base-version: '21.08'
```

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+      - type: file
+        path: ro.go.hmlendea.Duolingo.metainfo.xml
+      - type: file
+        path: ro.go.hmlendea.Duolingo.png
```
the icon being in the upstream directory you should add here. If it's not in the tarball, you should get it from the repository.

hfiguiere replied:
this file is no longer used and should removed or used

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+    sources:
+      - type: file
+        only-arches:
+          - x86_64
```
Also because if this you need to do this:

https://github.com/flathub/flathub/wiki/App-Maintenance#only-build-on-x86_64

hfiguiere replied:
this is sitll not done. Build will still fail.

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --socket=pulseaudio
```
you likely need `--device=dri` for GL rendering.

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+      - cp -r duolingo-desktop /app
+      - install -Dm 644 ro.go.hmlendea.Duolingo.png ${FLATPAK_DEST}/share/icons/hicolor/512x512/apps/${FLATPAK_ID}.png
+      - install -Dm 644 ro.go.hmlendea.Duolingo.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm 644 ro.go.hmlendea.Duolingo.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.appdata.xml
```
```suggestion
      - install -Dm 644 ro.go.hmlendea.Duolingo.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
```

hfiguiere commented:
bot, build ro.go.hmlendea.Duolingo

hfiguiere commented:
ideally the metainfo.xml (appstream) should be managed upstream.

hfiguiere commented on ro.go.hmlendea.Duolingo.metainfo.xml:
```diff
@@ -0,0 +1,25 @@
+  <name>Duolingo</name>
+  <project_license>GPL-3.0</project_license>
+  <developer_name>hmlendea</developer_name>
+  <summary>Unofficial desktop client for the Duolingo webapp</summary>
```
I think we need a clear disclaimer here that this is not an official client.

hmlendea commented:
> ideally the metainfo.xml (appstream) should be managed upstream.

That means I should move that file to the upstream repo and then source it like this?
```yaml
  sources:
      - type: file
        url: https://...etc...
```

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+    build-commands:
+      - unzip -d duolingo-desktop duolingo-desktop_*.zip
+      - cp -r duolingo-desktop /app
+      - install -Dm 644 icon.svg ${FLATPAK_DEST}/share/icons/hicolor/512x512/apps/duolingo.svg
```
```suggestion
      - install -Dm 644 icon.svg ${FLATPAK_DEST}/share/icons/hicolor/scalable/apps/duolingo.svg
```

hmlendea replied:
I've switched to the icon from upstream, which is a png, so I'll keep using the 512x512 dir

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+base: org.electronjs.Electron2.BaseApp
+base-version: '21.08'
+separate-locales: false
+command: duolingo-desktop
```
```suggestion
command: duolingo-desktop
rename-icon: duolingo
```

it's required to reame the icon including in the desktop file.

hmlendea replied:
The desktop file already contains `Icon=duolingo` so it's ok

hfiguiere replied:
This was ignored and this is why the build is failing.

hmlendea replied:
Disclaimer: I didn't mean to be offensive with the word "suggest". I used it because that's what the GitHub interface shows, and is also the preferred form at my workplace (request sounds a bit harsh)

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+      - type: file
+        path: ro.go.hmlendea.Duolingo.png
+      - type: file
+        url: https://design.duolingo.com/86230c9ad10d9f08b785.svg
```
you should add a sha256.

Also this look like "borrowed" from duolingo which isn't really fair use so it's probably not a good idea.

hmlendea replied:
As far as I understood from their page, that icon (along with others) are redistributable.

But just to be safe I'll just use the upstream app icon instead.

I've added the checksum as well.

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,53 @@
+    build-commands:
+      - unzip -d duolingo-desktop duolingo-desktop_*.zip
+      - cp -r duolingo-desktop /app
+      - install -Dm 644 icon.png ${FLATPAK_DEST}/share/icons/512x512/scalable/apps/duolingo.png
```
```suggestion
      - install -Dm 644 icon.png ${FLATPAK_DEST}/share/icons/512x512/apps/duolingo.png
```

hmlendea replied:
Oops, sorry about that.

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,53 @@
+            "_linux.zip") | .browser_download_url
+      - type: file
+        path: ro.go.hmlendea.Duolingo.metainfo.xml
+      - type: file
+        path: ro.go.hmlendea.Duolingo.png
```
```suggestion
```

hfiguiere commented:
bot, build ro.go.hmlendea.Duolingo

hfiguiere commented:
```
install: cannot stat 'duolingo-desktop/ro.go.hmlendea.Duolingo.desktop': No such file or directory
```

hfiguiere commented:
bot, build ro.go.hmlendea.Duolingo

hmlendea commented:
The build errored due to an XML syntax error. I've tested the file again using `xmlstarlint` and it passes validation

Off-topic: Any idea why the build process hangs for so long after an error, instead of failing right away?

hfiguiere commented:
bot, build ro.go.hmlendea.Duolingo

hfiguiere commented on ro.go.hmlendea.Duolingo.desktop:
```diff
@@ -0,0 +1,12 @@
+[Desktop Entry]
```
I do not understand while you brought it back?

hmlendea replied:
I've had trouble with the icon (but turned out it was something else) and the Exec (but I fixed it another way)

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+    build-commands:
+      - unzip -d duolingo-desktop duolingo-desktop_*.zip
+      - cp -r duolingo-desktop /app
+      - install -Dm 644 icon.png ${FLATPAK_DEST}/share/icons/512x512/apps/apps/${FLATPAK_ID}.png
```
```suggestion
      - install -Dm 644 icon.png ${FLATPAK_DEST}/share/icons/512x512/apps/${FLATPAK_ID}.png
```

hmlendea commented:
It should be fixed for good now

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+      - install -Dm 644 "duolingo-desktop/${FLATPAK_ID}.desktop" "${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop"
+      - install -Dm 644 "${FLATPAK_ID}.metainfo.xml" "${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml"
+      - install -Dm 755 "duolingo-desktop.sh" "/app/bin/duolingo-desktop"
+      - sed -i 's/^Exec=.*/Exec=duolingo-desktop/g' "${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop"
```
please use `desktop-file-edit` for these changes.

hfiguiere commented on ro.go.hmlendea.Duolingo.yaml:
```diff
@@ -0,0 +1,51 @@
+base-version: '21.08'
+separate-locales: false
+command: duolingo-desktop
+rename-icon: ro.go.hmlendea.Duolingo
```
```suggestion
```

the icon already has that name.

hmlendea replied:
It fails with `Error loading desktop file: Failed to find icon: Failed to find icon duolingo` if I remove that

hmlendea replied:
I could use `desktop-file-edit` to change it to `ro.go.hmlendea.Duolingo` in the desktop file but that's pretty redundant since the `rename-icon` seems to do the same thing (I think)

hfiguiere replied:
that's because when you installed the icon as `duolingo.png` and I told you to add the rename icon you ignored it. Now that you install it with the flatpak_id name, you have to fix the .desktop file as I mentioned.

You still don't.

hmlendea replied:
I didn't ignore anything. I tried to make the changes as I understood them.
Hopefully this latest commit is.

hfiguiere replied:
you still did.

See THIS: https://github.com/flathub/flathub/pull/2926#discussion_r828428799

hmlendea replied:
Like this then? https://github.com/flathub/flathub/pull/2926/commits/ba5b9c9817a3048baf8263f6a71172c246bba67b

hfiguiere commented:
bot, build ro.go.hmlendea.Duolingo

hmlendea commented:
Installed locally using the provided command and it works as expected

hmlendea commented:
So is it ok like this? Or is there something that still needs to change?

hmlendea commented:
If not, can this be merged then?

barthalion commented:
My primary concern is the name; it's not really Duolingo, which I'm sure is a registered trademark. As it's just a webview, no amount of "this is fair use" arguments will save us from Duolingo lawyers contacting us. If it gets renamed to something else, and the description is absolutely clear what it is, it's good to be merged.

hmlendea commented:
Understood.

What do you think would be a good and safe name to use? Can something like Duolingo Desktop work or Duolingo Electron (akin to how GeForce NOW Electron is named)?

I've updated the description to position the "unofficial" notice as the first thing.

hmlendea commented:
I've went ahead and renamed it to `DL: language lessons` as it seems like it is the safest approach _(since it contains neither Duolingo, nor Duo, in the name)_.
I've also replaced the icon and renamed the app ID as well _(ro.go.hmlendea.DL-Desktop)_

I've tested the build locally and everything seems alright

barthalion commented:
bot, build ro.go.hmlendea.DL-Desktop

hmlendea commented:
I've installed the version from the bot's build also, and this one is working as well.

barthalion commented:
/merge
